### PR TITLE
Laghos GPU addons [laghos]

### PR DESCRIPTION
**Pull request by camierjs**
This PR brings:

- a couple of optimizations from @pwang234 (`SmemPAMassApply3D` & `ElementRestriction::Mult`),
- few new `QuadratureInterpolator` tensor functions:
   - `Values`: interpolate the values of the E-vector at quadrature points,
   - `Derivatives`: interpolate the derivatives of the E-vector,
   - `PhysDerivatives`: interpolate the derivatives in physical space of the E-vector,
- a way to propose small linear algebra code useful on the GPU now in `linalg/blas.hpp`.

---
Before:
[plot2_MFEM_cuda_bp1_linux_gcc_N001_std.pdf](https://github.com/mfem/mfem/files/3990866/plot2_MFEM_cuda_bp1_linux_gcc_N001_std.pdf)

After:
[plot2_MFEM_cuda_bp1_linux_gcc_N001_all.pdf](https://github.com/mfem/mfem/files/3990869/plot2_MFEM_cuda_bp1_linux_gcc_N001_all.pdf)

---
- [x] `NVTX` cleanup
- [x] move code into `linalg/blas.hpp` 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1208](https://github.com/mfem/mfem/pull/1208) | @camierjs | @tzanio | @artv3 + @v-dobrev | 01/24/20 | 02/26/20 | 03/10/20 | |

### CODE REVIEW

**Comment by artv3** [comment]
Is there a branch of Laghos where this is being tested with?

**Comment by camierjs** [comment]
> Is there a branch of Laghos where this is being tested with?

It's the [okina](https://github.com/CEED/Laghos/tree/okina) branch of Laghos, yes.

**Comment by artv3** [comment]
@camierjs , perhaps the Laghos okina branch needs to be updated.
 I tried building with this branch and hit some errors: 
```
laghos_solver.cpp(1076): error: class "mfem::DenseMatrix" has no member "symmetrize"
laghos_solver.cpp(1091): error: class "mfem::Vector" has no member "norml2"
....
```
The methods should be capitalized in Laghos.

**Comment by camierjs** [comment]
Ok, thanks @artv3, I updated Laghos:okina.

File: fem/fespace.hpp [hunk 1]
```diff
@@ -983,6 +983,14 @@ class QuadratureInterpolator
...
              Vector &q_val, Vector &q_der, Vector &q_det) const;
 
+   /// Interpolate the values of the E-vector @a e_vec at quadrature points.
+   /// This computation uses tensor product structure.
+   void Values(const Vector &e_vec, Vector &q_val) const;
+
+   /// Interpolate the derivatives of the E-vector @a e_vec at quadrature points.
+   /// This computation uses tensor product structure.
+   void Derivatives(const Vector &e_vec, Vector &q_der) const;
+
```
Comment (artv3) [inline fem/fespace.hpp]: It looks like this computes derivatives with respect to reference space. Could we add a method that maps it back to physical space? I think this is generic enough that we should be able to calculate derivatives at dofs (as I need) or quadrature points.

Comment (camierjs) [reply fem/fespace.hpp]: `done` via PR #1270 , thanks @artv3

**Comment by artv3** [comment]
Looks good to me! Only thought I have is should we change the blas namespace to something else?

**Comment by tzanio** [comment]
@v-dobrev, the review on this PR is due this week

**Comment by tzanio** [comment]
Adding a few extra days for approval...

**Comment by tzanio** [comment]
Can you please merge master and resolve the conflicts?

File: fem/fespace.cpp [hunk 2]
```diff
@@ -3273,4 +3273,888 @@ void QuadratureInterpolator::MultTranspose(
    MFEM_ABORT("this method is not implemented yet");
 }
 
+template<int T_D1D =0, int T_Q1D =0, int T_NBZ =0>
+static void D2QValues2D(const int NE,
+                        const Array<double> &b_,
```
Comment (tzanio) [inline fem/fespace.cpp]: Should this go into a different file now?

File: fem/fespace.cpp [hunk 3]
```diff
@@ -3249,6 +3248,7 @@ void QuadratureInterpolator::Mult(
             // Q4
             case 125125: eval_func = &Eval3D<3,125,125>; break;
             case 125216: eval_func = &Eval3D<3,125,216>; break;
+               //case 125512: eval_func = &Eval3D<3,125,512>; break;
```
Comment (tzanio) [inline fem/fespace.cpp]: Remove

File: fem/fespace.cpp (hunk 2 above)
Comment (camierjs) [reply fem/fespace.cpp]: Yes, it is in `fem/quadinterpolator.cpp` now

File: fem/fespace.cpp (hunk 3 above)
Comment (camierjs) [reply fem/fespace.cpp]: I don't see it anymore.

Comment (camierjs) [reply fem/fespace.cpp]: It must have been removed before the merge with master.

File: fem/restriction.cpp [hunk 4]
```diff
@@ -152,24 +152,18 @@ void ElementRestriction::Mult(const Vector& x, Vector& y) const
...
-         const double dofValue = d_x(t?c:i,t?i:c);
-         for (int j = offset; j < nextOffset; ++j)
-         {
-            const bool positive = (d_indices[j] >= 0);  // otherwise, sign must be flipped
-            const int idx_j = positive ? d_indices[j] : -1 - d_indices[j];
-            d_y(idx_j % nd, c, idx_j / nd) = positive ? dofValue : -dofValue;
-         }
+         const double dofValue = d_x(t?c:j, t?j:c);
+         d_y(i % nd, c, i / nd) = plus ? dofValue : -dofValue;
       }
```
Comment (tzanio) [inline fem/restriction.cpp]: Is the logic here still the same?

File: general/forall.hpp [hunk 5]
```diff
@@ -31,6 +31,25 @@
...
+#define MFEM_PRAGMA(X) _Pragma(#X)
+// MFEM_UNROLL pragma macro
+#if defined(MFEM_USE_CUDA) || \
+    (defined(__clang__) && ((__clang_major__ > 3) || \
+                            (__clang_major__ == 3 && __clang_minor__ >= 7)))
+#define MFEM_UNROLL(N) MFEM_PRAGMA(unroll N)
+#elif defined(__xlC__)
+#define MFEM_UNROLL(N) MFEM_PRAGMA(unroll(N))
+#elif defined(__ICC)
+// Using "MFEM_PRAGMA(unroll(N))" results in compiler remarks, so don't use it
```
[the remark this answers is not in the document]
Comment (camierjs) [reply general/forall.hpp]: Right, that still needs to be done

File: fem/restriction.cpp (hunk 4 above)
Comment (camierjs) [reply fem/restriction.cpp]: We had to solve the conflict, but it should be fine.

File: linalg/densemat.cpp [hunk 6]
```diff
@@ -172,7 +172,7 @@ const double &DenseMatrix::Elem(int i, int j) const
 
 void DenseMatrix::Mult(const double *x, double *y) const
 {
-   kernels::MultV(height, width, Data(), x, y);
+   kernels::Mult(height, width, Data(), x, y);
```
Comment (tzanio) [inline linalg/densemat.cpp]: @camierjs, fyi I changed the name of this function.

File: linalg/kernels.hpp [hunk 7]
```diff
@@ -65,155 +68,183 @@ double Norml2(const int size, const T *data)
...
+template<int dim>
+void CalcEigenvalues(const double *data, double *lambda, double *vec);
+
+/// Return the i'th singular value of the matrix of size dim with given @a data.
+template<int dim>
+double CalcSingularvalue(const double *data, const int i);
+
+
+// Utility functions for CalcEigenvalues and CalcSingularvalue
+namespace internal
```
Comment (tzanio) [inline linalg/kernels.hpp]: New namespace `mfem::kernels::internal` for the utility functions

**Comment by tzanio** [comment]
@camierjs, I believe the changes in 9b19b8c are correct, but may be worth double-checking in Laghos, etc.

File: linalg/kernels.hpp (hunk 7 above)
Comment (camierjs) [reply linalg/kernels.hpp]: Are you sure about the comma in `EXCLUDE_SYMBOLS`?
The documentation script now fails in an `internal` class of the `mem_manager`.

Comment (tzanio) [reply linalg/kernels.hpp]: Let me take a look...

Comment (tzanio) [reply linalg/kernels.hpp]: You were right, fixed

**Comment by tzanio** [comment]
Merged in `next` for testing ...

**Comment by tzanio** [comment]
There are a number of regression test differences from the 03/08/20 runs on [tux426](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/a795aff2d45bf3223d7a647875d5955ad5eecba5) and [tux429](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/3a5273702c34f477218b14655c92468e4c262cef) with this branch.

On `tux429` the following runs have round-off level errors:
```
ex3 -m ../data/beam-hex.mesh
mpirun -np 4 ex3p -m ../data/star.mesh -pa -d cuda -no-vis
mpirun -np 4 ex3p -m ../data/star.mesh -pa -d raja-cuda -no-vis
mpirun -np 4 ex3p -m ../data/star.mesh -pa -d raja-omp -no-vis
mpirun -np 4 ex3p -m ../data/beam-hex.mesh -pa -d cuda -no-vis
```

On `tux426` the following runs have round-off level errors:
```
ex3 -m ../data/beam-hex.mesh -o 2 -pa -no-vis
ex3 -m ../data/square-disc-nurbs.mesh -no-vis
ex3 -m ../data/amr-hex.mesh -no-vis
ex10 -m ../data/beam-quad.mesh -s 3 -r 2 -o 2 -dt 3 -no-vis
ex10 -m ../data/beam-hex.mesh -s 2 -r 1 -o 2 -dt 3 -no-vis
ex10 -m ../data/beam-wedge.mesh -s 2 -r 1 -o 2 -dt 3 -no-vis
ex10 -m ../data/beam-hex.mesh -s 14 -r 1 -o 2 -dt 0.05 -vs 20 -no-vis
ex10 -m ../data/beam-quad-amr.mesh -s 3 -r 2 -o 2 -dt 3 -no-vis
ex17 -m ../data/beam-hex.mesh -no-vis
ex17 -m ../data/beam-wedge.mesh -no-vis
ex17 -m ../data/beam-quad.mesh -r 2 -o 3 -no-vis
ex17 -m ../data/beam-hex.mesh -r 2 -o 2 -no-vis
ex21 -m ../data/beam-quad.mesh -o 3 -no-vis
ex21 -m ../data/beam-hex.mesh -no-vis
ex22 -m ../data/inline-quad.mesh -o 3 -p 1 -no-vis
ex24 -m ../data/fichera-q3.mesh -no-vis
... (many more parallel runs)
```

Is this expected or is it likely that there is a bug in the dense matrix functions?

**Comment by camierjs** [comment]
The `PA` `mass` kernel has changed, as well as the restriction, so it might impact at round-off level all examples/apps using it, yes.

**Comment by tzanio** [comment]
On further investigation the issues seems to be in 4eca3f0 (the results from the previous commit 5579f0b, at least on `ex3 -m ../data/beam-hex.mesh -o 2 -pa -no-vis` appear to be the same).

@v-dobrev, are these due to these changes: https://github.com/mfem/mfem/commit/4eca3f0ba8ee908cdf88a5ca936c66c987ce35f6#diff-8c02460d99a43b0e608d4e0d8e1c3a40, and if so do we really expect them to affect so many examples, many of them not using PA?

**Comment by tzanio** [comment]
> Could the namespace be ::kernel_utils?

We had a long discussion about the name. What's the concern with `mfem::kernels`?

**Comment by artv3** [comment]
> > Could the namespace be ::kernel_utils?
> 
> We had a long discussion about the name. What's the concern with `mfem::kernels`?

I was thinking the functions are more like utilities for kernels, but the namespace is fine with me.

**Comment by tzanio** [comment]
We are still open for suggestions... I was not completely happy with `kernels` :)

Some other things we considered: `blas`, `linalg`, `small`, `inlined` or `inlines`, ...

These free functions are the implementations of dense linear algebra algorithms for "small" matrices and vectors that can be directly inlined in CPU functions and/or device kernels. It will be good if somehow the namespace name reflects this :)

**Comment by camierjs** [comment]
I still have an issue: 
`c++: error: unrecognized command line option ‘-march=native’`
on `Lassen` with `miniapps/performance/CMakeFiles/performance_ex1.dir/ex1.cpp.o`

**Comment by tzanio** [comment]
This is governed by the `MFEM_PERF_SW` variable in `miniapps/performance/makefile'. It probably has to be updated for Lassen.

See also, https://github.com/CEED/benchmarks/blob/master/machine-configs/lassen.sh

**Comment by camierjs** [comment]
Related to [#219](https://github.com/mfem/mfem/pull/219)

**Pull request merged**